receiver/snmpreceiver/config.go

```diff
@@ -14,17 +14,578 @@
+	}
+}
+
+// validateAuth validates the AuthType and AuthPassword
+func validateAuth(cfg *Config) error {
+	var combinedErr error
+
+	// Ensure valid auth password
+	if cfg.AuthPassword == "" {
+		combinedErr = multierr.Append(combinedErr, errEmptyAuthPassword)
+	}
+
+	// Ensure valid auth type
+	if cfg.AuthType == "" {
+		return multierr.Append(combinedErr, errEmptyAuthType)
+	}
+
+	switch strings.ToUpper(cfg.AuthType) {
+	case "MD5", "SHA", "SHA224", "SHA256", "SHA384", "SHA512": // ok
+	default:
+		combinedErr = multierr.Append(combinedErr, errBadAuthType)
+	}
+
+	return combinedErr
+}
+
+// validatePrivacy validates the PrivacyType and PrivacyPassword
+func validatePrivacy(cfg *Config) error {
+	var combinedErr error
+
+	// Ensure valid privacy password
+	if cfg.PrivacyPassword == "" {
+		combinedErr = multierr.Append(combinedErr, errEmptyPrivacyPassword)
+	}
+
+	// Ensure valid privacy type
+	if cfg.PrivacyType == "" {
+		return multierr.Append(combinedErr, errEmptyPrivacyType)
+	}
+
+	switch strings.ToUpper(cfg.PrivacyType) {
+	case "DES", "AES", "AES192", "AES192C", "AES256", "AES256C": // ok
+	default:
+		combinedErr = multierr.Append(combinedErr, errBadPrivacyType)
+	}
+
+	return combinedErr
+}
+
+// validateMetricConfigs validates all MetricConfigs, AttributeConfigs, and ResourceAttributeConfigs
+func validateMetricConfigs(cfg *Config) error {
+	var combinedErr error
+
+	// Validate the Attribute and ResourceAttribute configs up front
+	combinedErr = multierr.Append(combinedErr, validateAttributeConfigs(cfg))
+	combinedErr = multierr.Append(combinedErr, validateResourceAttributeConfigs(cfg))
+
+	// Ensure there is at least one MetricConfig
+	metrics := cfg.Metrics
+	if len(metrics) == 0 {
+		return multierr.Append(combinedErr, errMetricRequired)
+	}
+
+	// Make sure each MetricConfig has valid info
+	for metricName, metricCfg := range metrics {
+		if metricCfg.Unit == "" {
+			combinedErr = multierr.Append(combinedErr, fmt.Errorf(errMsgMetricNoUnit, metricName))
+		}
+
+		if metricCfg.Gauge == nil && metricCfg.Sum == nil {
+			combinedErr = multierr.Append(combinedErr, fmt.Errorf(errMsgMetricNoGaugeOrSum, metricName))
+		}
+
+		if len(metricCfg.ScalarOIDs) == 0 && len(metricCfg.ColumnOIDs) == 0 {
+			combinedErr = multierr.Append(combinedErr, fmt.Errorf(errMsgMetricNoOIDs, metricName))
+		}
+
+		if metricCfg.Gauge != nil {
+			combinedErr = multierr.Append(combinedErr, validateGauge(metricName, metricCfg.Gauge))
+		}
+
+		if metricCfg.Sum != nil {
+			combinedErr = multierr.Append(combinedErr, validateSum(metricName, metricCfg.Sum))
+		}
+
+		if len(metricCfg.ScalarOIDs) > 0 {
+			for _, scalarOID := range metricCfg.ScalarOIDs {
+				combinedErr = multierr.Append(combinedErr, validateScalarOID(metricName, scalarOID, cfg))
+			}
+		}
+
+		if len(metricCfg.ColumnOIDs) > 0 {
+			for _, columnOID := range metricCfg.ColumnOIDs {
+				combinedErr = multierr.Append(combinedErr, validateColumnOID(metricName, columnOID, cfg))
+			}
+		}
```

Comment:
Typically the length is not checked directly unless necessary, since the loop will just not execute.

```suggestion
		for _, scalarOID := range metricCfg.ScalarOIDs {
			combinedErr = multierr.Append(combinedErr, validateScalarOID(metricName, scalarOID, cfg))
		}

		for _, columnOID := range metricCfg.ColumnOIDs {
			combinedErr = multierr.Append(combinedErr, validateColumnOID(metricName, columnOID, cfg))
		}
```

Reply:
@djaglowski consider this a brain blip haha. maybe at one point i had logic in here other than the for loop? i'll make this change